### ISSUE TITLE
Remove the dependency on boto

### ISSUE THREAD

**Issue by GaelVaroquaux** [nistats/nistats#345]
The dependency on boto is a blocker for integrating in nilearn. It should be removed from nistats.

It should be possible to implement the corresponding functionality (data download) with the Python standard library.

**Comment by kchawla-pi** [comment]
I am looking into this now. Boto creates these urls to download from using some base path and I am guessing crawls through the server, finding every file that matches the filter to create a list of URLs to download from. The url list is  411759 entries.
I don't know where to begin to generate so many URLs using standar python.

**Comment by effigies** [comment]
@kchawla-pi Would it make sense to have it as an optional dependency? You could add it somewhere in `extras_require` for `setuptools` and then have any internal logic use [nibabel.optpkg](https://github.com/nipy/nibabel/blob/master/nibabel/optpkg.py):

```Python
from nibabel.optpkg import optional_package
boto3, has_boto3, setup_module = optional_package('boto3')
```

If it's missing, attempts to use the `boto3` "module" will result in an informative error asking users to install it. And you can condition some things on `has_boto3`.

**Comment by kchawla-pi** [comment]
It's already optional, only imported when a specific dataset fetcher is called. Currently it raises an error, instead of asking nicely, but that is easily fixed.

**Comment by GaelVaroquaux** [comment]
But that means that examples cannot run without boto, which makes
teaching harder.

I don't understand why it should be mandatory. There must be a way to
query the server to get the file list.

**Comment by jeromedockes** [comment]
> The dependency on boto is a blocker for integrating in nilearn. It should be removed from nistats.
> 

 +1

> It should be possible to implement the corresponding functionality (data download) with the Python standard library.

yes. although at some point we could consider adding `requests` as a dependency, since it is ubiquitous and much easier to use than the standard library modules

**Comment by GaelVaroquaux** [comment]
> although at some point we could consider adding requests as a
> dependency, since it is ubiquitous and much easier to use than the
> standard library modules

Yes.

**Comment by kchawla-pi** [comment]
>  although at some point we could consider adding requests as a
>  dependency, since it is ubiquitous and much easier to use than the
>  standard library modules

> Yes.

+1
Then we should slowly and systematically replace existing solutions with requests, else we'll be trading one dependency for another if we only use it once.

**Comment by kchawla-pi** [comment]
A workaround for this issue can be that we upload the urls (currrently saved as a json file by the fetcher to osf.io and the fetcher downloads that, and uses that information to download the dataset itself.
It should not require much work.

**Comment by kchawla-pi** [comment]
Implementing a PR based on this idea.

**Comment by GaelVaroquaux** [comment]
Sounds good.

Make sure to add a script to generate these URLs somewhere. This script
should be run by maintainers, and not users, and can have boto as a
dependency.

**Comment by kchawla-pi** [comment]
Good idea. I'll keep the original index downloader function. We can rerun that as and when necessary and do a new manual upload.
